proposed/2021/DotNetAudit.md

```diff
@@ -0,0 +1,296 @@
+
+Run 'dotnet audit fix' to fix them.
+
+warning: Outdated packages found!
+[net5.0]:
+Top-level Package						            Resolved	Latest
+> anurse.testing.TestDeprecatedPackage	            1.0.0		2.0.0
+> UmbracoForms							            8.4.1	    8.7.1
+
+Found 2 top-level outdated package(s) in 36 scanned packages.
+
+Run 'dotnet audit fix' to fix them.
+```
+
+##### Endpoints
+
+NuGet will use existing endpoints to optimize the speed of audit results.
+
+- [Deprecation](https://docs.microsoft.com/en-us/nuget/api/registration-base-url-resource#package-deprecation)
+- [Vulnerability](https://docs.microsoft.com/en-us/nuget/api/registration-base-url-resource#vulnerabilities)
+- Outdated - No existing endpoint, will need to call a source.
+
+##### dotnet audit --json
+
+To get a detailed audit report in a JSON format.
+
+```
+{
+  "audit": {
+    "vulnerabilities": {
+      "low": "0",
+      "moderate": "2",
+      "high": "0",
+      "critical": "0",
+      "dependencies": [
+        {
+          "name": "UmbracoForms",
+          "requestedVersion": "8.4.1",
+          "resolvedVersion": "8.4.1",
+          "vulnerabilitySeverity": "Moderate",
+          "advisoryUrl": "https://github.com/advisories/GHSA-8m73-w2r2-6xxj",
+          "transitiveDependencies": [
+            {
+              "name": "Microsoft.Data.OData",
+              "resolvedVersion": "5.2.0",
+              "severity": "Moderate",
+              "advisoryUrl": "https://github.com/advisories/GHSA-mv2r-q4g5-j8q5"
+            }
+          ]
+        }
+      ]
+    },
+    "deprecations": {
+      "legacy": "1",
+      "critical-bugs": "0",
+      "other": "0",
+      "dependencies": [
+        {
+          "name": "anurse.testing.TestDeprecatedPackage",
+          "requestedVersion": "1.0.0",
+          "resolvedVersion": "1.0.0",
+          "reason": "Legacy",
+          "alternativeDependencies": [
+            {
+              "name": "Microsoft.AspNetCore.MVC",
+              "version": "0.0.0"
+            }
+          ]
+        }
+      ]
+    },
+    "outdated": {
+      "packages": "2",
+      "dependencies": [
+        {
+          "name": "anurse.testing.TestDeprecatedPackage",
+          "resolvedVersion": "1.0.0",
+          "latestVersion": "2.0.0"
+        },
+        {
+          "name": "UmbracoForms",
+          "resolvedVersion": "8.4.1",
+          "latestVersion": "8.7.1"
+        }
+      ]
+    },
+    "scannedDependencies": "36"
+  }
+}
+```
+
+##### dotnet audit Exit Codes
+
+- 0 - The command will exit with a 0 exit code if no vulnerabilities, deprecations, or outdated packages were found.
+- 1 - The command will exit with a 1 exit code if a vulnerability, deprecation, or outdated package was found.
+- 2 - The command will exit with a 2 exit code if it unexpectedly failed.
+- 3 - The command will exit with a 3 exit code if an unsupported project is detected.
+
+#### dotnet audit fix
```

Comment:
I'm sure we have more duplicates of the same request, but `dotnet audit fix` just isn't going to work for a whole lot of customers, including ASP.NET Core customers using an LTS (not latest) version: https://github.com/NuGet/Home/issues/11457

Unless we're planning on implementing `dotnet audit fix` after implementing the above feature, I think this spec should document this known limitation.

Reply:
This is quite a hard challenge indeed. I think we can do our best job designing this feature for the majority of packages that aren't tied directly to a .NET version in NuGet. Even then, there is value for when vulnerabilities or deprecations are found.

i.e. https://www.nuget.org/packages/System.DirectoryServices.Protocols/ may have vulnerabilities in NET5 & prompt a user to migrate to NET6 or higher. I don't think this tool will be the panacea of modernization, but it can help bring awareness to the benefit of doing so or even point users to the respective modernization tools if needed. 

If we do the work in the issue that allows for such functionality, we can port it over to this tool / try our best to detect it from the package lists to be smarter in these cases.

I'll add similar notes in the proposal.